File: typescript/src/json.ts
/**
 * JSON serialization utilities for deterministic JSON encoding.
 * Provides helpers for encoding Cramberry messages to/from JSON with
 * blockchain-grade determinism.
 */

/**
 * Formats a bigint as a string for JSON encoding.
 * All integers are encoded as strings to prevent JavaScript precision loss.
 */
export function formatBigIntToString(value: bigint): string {
  return value.toString();
}

/**
 * Formats a number (int32, uint32) as a string for JSON encoding.
 */
export function formatNumberToString(value: number): string {
  return Math.floor(value).toString();
}

/**
 * Formats a float32 with 9 significant digits for deterministic JSON.
 * Matches the output of Go's strconv.FormatFloat(v, 'g', 9, 32).
 * Throws error if the value is NaN or Infinity.
 */
export function formatFloat32(value: number): string {
  if (isNaN(value)) {
    throw new Error('cannot encode NaN to JSON');
  }
  if (!isFinite(value)) {
    throw new Error('cannot encode Infinity to JSON');
  }
  if (value === 0) {
    return '0';
  }
  // Narrow to f32 precision so the formatted bits match what Go would
  // produce when given a float32. Math.fround returns the f64 closest to
  // the f32 quantization of value.
  return formatGoG(Math.fround(value), 9);
}

/**
 * Formats a float64 with 17 significant digits for deterministic JSON.
 * Matches the output of Go's strconv.FormatFloat(v, 'g', 17, 64).
 * Throws error if the value is NaN or Infinity.
 */
export function formatFloat64(value: number): string {
  if (isNaN(value)) {
    throw new Error('cannot encode NaN to JSON');
  }
  if (!isFinite(value)) {
    throw new Error('cannot encode Infinity to JSON');
  }
  if (value === 0) {
    return '0';
  }
  return formatGoG(value, 17);
}

/**
 * Formats a non-zero finite number using Go's strconv 'g' verb with the
 * given precision (number of significant digits).
 *
 *   - Decimal form when -4 <= exponent < precision.
 *   - Otherwise scientific: `mantissa "e" sign digits` with the exponent
 *     always signed and at least two digits.
 *   - Trailing zeros in the fractional part are stripped.
 */
function formatGoG(value: number, precision: number): string {
  // toExponential(p) gives p digits after the decimal — i.e. p+1 significant
  // digits total — using IEEE round-half-to-even, the same rule Go uses.
  const exp = value.toExponential(precision - 1);
  const match = /^(-?\d+(?:\.\d+)?)e([+-]?\d+)$/.exec(exp);
  if (!match) {
    throw new Error('toExponential produced unexpected format: ' + exp);
  }
  let mantissa = match[1];
  const exponent = parseInt(match[2], 10);

  // Strip trailing zeros from the fractional part of the mantissa.
  if (mantissa.includes('.')) {
    mantissa = mantissa.replace(/0+$/, '');
    if (mantissa.endsWith('.')) mantissa = mantissa.slice(0, -1);
  }

  if (exponent >= -4 && exponent < precision) {
    return decimalFromMantissaExp(mantissa, exponent);
  }
  const sign = exponent >= 0 ? '+' : '-';
  const absExp = Math.abs(exponent);
  const expStr = absExp < 10 ? '0' + absExp : String(absExp);
  return mantissa + 'e' + sign + expStr;
}

/**
 * Renders mantissa * 10^exp in plain decimal form, stripping trailing zeros
 * from any fractional part.
 */
function decimalFromMantissaExp(mantissa: string, exp: number): string {
  let sign = '';
  if (mantissa.startsWith('-')) {
    sign = '-';
    mantissa = mantissa.slice(1);
  }
  const dotIdx = mantissa.indexOf('.');
  const digits =
    dotIdx < 0 ? mantissa : mantissa.slice(0, dotIdx) + mantissa.slice(dotIdx + 1);
  const integerLen = dotIdx < 0 ? mantissa.length : dotIdx;
  const newIntegerLen = integerLen + exp;

  if (newIntegerLen <= 0) {
    const padding = '0'.repeat(-newIntegerLen);
    let result = '0.' + padding + digits;
    result = result.replace(/0+$/, '');
    if (result.endsWith('.')) result = result.slice(0, -1);
    return sign + result;
  }
  if (newIntegerLen >= digits.length) {
    const padding = '0'.repeat(newIntegerLen - digits.length);
    return sign + digits + padding;
  }
  const intPart = digits.slice(0, newIntegerLen);
  const fracPart = digits.slice(newIntegerLen).replace(/0+$/, '');
  if (fracPart === '') return sign + intPart;
  return sign + intPart + '.' + fracPart;
}

/**
 * Validates that a float is encodable to JSON.
 */
export function validateFloat(value: number): void {
  if (isNaN(value)) {
    throw new Error('cannot encode NaN to JSON');
  }
  if (!isFinite(value)) {
    throw new Error('cannot encode Infinity to JSON');
  }
}

/**
 * Encodes a Uint8Array to base64 string (RFC 4648 standard encoding).
 */
export function encodeBase64(data: Uint8Array): string {
  // Node.js
  if (typeof Buffer !== 'undefined') {
    return Buffer.from(data).toString('base64');
  }
  // Browser
  const binary = Array.from(data, byte => String.fromCharCode(byte)).join('');
  return btoa(binary);
}

/**
 * Decodes a base64 string to Uint8Array.
 */
export function decodeBase64(str: string): Uint8Array {
  // Node.js
  if (typeof Buffer !== 'undefined') {
    return new Uint8Array(Buffer.from(str, 'base64'));
  }
  // Browser
  const binary = atob(str);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

/**
 * Parses a bigint from a JSON string or number.
 */
export function parseBigIntFromJSON(value: string | number): bigint {
  if (typeof value === 'string') {
    return BigInt(value);
  }
  // Accept numeric values but lose precision for large numbers
  return BigInt(Math.floor(value));
}

/**
 * Parses a strict integer from a JSON string or number.
 *
 * Rejects malformed inputs:
 *   - "123.5" / "12abc" / "" / " 12" → throws Error.
 *   - non-integer numbers like 12.5 → throws Error.
 * This matches Go's strconv.ParseInt which rejects all of the above.
 * The previous implementation used `parseInt(value, 10)`, which silently
 * truncated "123.5" to 123 and "12abc" to 12.
 */
export function parseNumberFromJSON(value: string | number): number {
  if (typeof value === 'string') {
    if (!/^-?\d+$/.test(value)) {
      throw new Error(`invalid integer: ${JSON.stringify(value)}`);
    }
    const n = Number(value);
    if (!Number.isFinite(n) || !Number.isSafeInteger(n)) {
      throw new Error(`integer out of safe range: ${value}`);
    }
    return n;
  }
  if (!Number.isFinite(value) || !Number.isInteger(value)) {
    throw new Error(`expected integer, got ${value}`);
  }
  return value;
}

const __sortKeyEncoder = new TextEncoder();

/**
 * Compares two strings by raw UTF-8 byte order, matching Go's `sort.Strings`.
 *
 * The default JS string comparator (`<`/`>`) uses UTF-16 code unit order,
 * which agrees with UTF-8 for BMP characters but disagrees for code points
 * `>= U+10000` (which encode as a surrogate pair `0xD800..0xDFFF` in UTF-16
 * but as bytes starting with `0xF0` in UTF-8). Using this comparator ensures
 * non-BMP keys sort the same way they do on the Go side.
 */
export function compareUtf8(a: string, b: string): number {
  const ab = __sortKeyEncoder.encode(a);
  const bb = __sortKeyEncoder.encode(b);
  const n = Math.min(ab.length, bb.length);
  for (let i = 0; i < n; i++) {
    if (ab[i] !== bb[i]) return ab[i] - bb[i];
  }
  return ab.length - bb.length;
}

/**
 * Sorts map keys lexicographically by UTF-8 byte order.
 * This ensures deterministic JSON output for maps.
 */
export function sortMapKeysLexicographic(keys: string[]): string[] {
  return [...keys].sort(compareUtf8);
}

/**
 * Returns a JSON string literal — the escaped contents *plus* surrounding
 * double quotes. Equivalent to Go's `cramberry.EscapeJSONString`. Use this
 * when you need a complete JSON string token to splice into a JSON document.
 */
export function escapeJSONString(str: string): string {
  return JSON.stringify(str);
}

/**
 * JSONWriter helps build JSON strings efficiently.
 */
export class JSONWriter {
  private parts: string[] = [];

  writeString(s: string): void {
    this.parts.push(s);
  }

  toString(): string {
    return this.parts.join('');
  }

  reset(): void {
    this.parts = [];
  }
}

/**
 * JSONReader helps parse JSON with validation.
 */
export class JSONReader {
  private data: any;

  constructor(json: string) {
    this.data = JSON.parse(json);
  }

  readObject(): Record<string, any> {
    if (typeof this.data !== 'object' || this.data === null || Array.isArray(this.data)) {
      throw new Error('expected JSON object');
    }
    return this.data;
  }

  getData(): any {
    return this.data;
  }
}

File: rust/src/json.rs
//! JSON serialization utilities for deterministic JSON encoding.
//! Provides helpers for encoding Cramberry messages to/from JSON with
//! blockchain-grade determinism.

use base64::{engine::general_purpose, Engine as _};

/// Formats an i64 as a string for JSON encoding.
/// All integers are encoded as strings to prevent JavaScript precision loss.
pub fn format_i64_to_string(value: i64) -> String {
    value.to_string()
}

/// Formats a u64 as a string for JSON encoding.
pub fn format_u64_to_string(value: u64) -> String {
    value.to_string()
}

/// Formats an i32 as a string for JSON encoding.
pub fn format_i32_to_string(value: i32) -> String {
    value.to_string()
}

/// Formats a u32 as a string for JSON encoding.
pub fn format_u32_to_string(value: u32) -> String {
    value.to_string()
}

/// Formats an i16 as a string for JSON encoding.
pub fn format_i16_to_string(value: i16) -> String {
    value.to_string()
}

/// Formats a u16 as a string for JSON encoding.
pub fn format_u16_to_string(value: u16) -> String {
    value.to_string()
}

/// Formats an i8 as a string for JSON encoding.
pub fn format_i8_to_string(value: i8) -> String {
    value.to_string()
}

/// Formats a u8 as a string for JSON encoding.
pub fn format_u8_to_string(value: u8) -> String {
    value.to_string()
}

/// Formats a f32 with 9 significant digits for deterministic JSON.
/// Matches the output of Go's `strconv.FormatFloat(v, 'g', 9, 32)`.
/// Returns error if the value is NaN or Infinity.
pub fn format_f32(value: f32) -> Result<String, String> {
    if value.is_nan() {
        return Err("cannot encode NaN to JSON".to_string());
    }
    if value.is_infinite() {
        return Err("cannot encode Infinity to JSON".to_string());
    }
    if value == 0.0 {
        return Ok("0".to_string());
    }
    // Widening to f64 preserves the f32 bit pattern exactly, so the 9-digit
    // round in format_go_g produces the same output as Go's '%.9g' on a f32.
    Ok(format_go_g(value as f64, 9))
}

/// Formats a f64 with 17 significant digits for deterministic JSON.
/// Matches the output of Go's `strconv.FormatFloat(v, 'g', 17, 64)`.
/// Returns error if the value is NaN or Infinity.
pub fn format_f64(value: f64) -> Result<String, String> {
    if value.is_nan() {
        return Err("cannot encode NaN to JSON".to_string());
    }
    if value.is_infinite() {
        return Err("cannot encode Infinity to JSON".to_string());
    }
    if value == 0.0 {
        return Ok("0".to_string());
    }
    Ok(format_go_g(value, 17))
}

/// Formats a non-zero finite `f64` using Go's `strconv` 'g' verb with the
/// given precision (number of significant digits).
///
///   - Decimal form when `-4 <= exponent < precision`.
///   - Otherwise scientific: `mantissa "e" sign digits` with the exponent
///     always signed and at least two digits.
///   - Trailing zeros in the fractional part are stripped.
fn format_go_g(value: f64, precision: usize) -> String {
    // {:.p$e} formats with `p` digits after the decimal point — i.e. p+1
    // significant digits total — using IEEE round-half-to-even, matching Go.
    let raw = format!("{:.*e}", precision - 1, value);
    let (mantissa_raw, exp_str) = match raw.find('e') {
        Some(idx) => (&raw[..idx], &raw[idx + 1..]),
        None => return raw,
    };
    let exponent: i32 = exp_str.parse().unwrap_or(0);

    let mantissa = strip_fractional_trailing_zeros(mantissa_raw);

    if exponent >= -4 && exponent < precision as i32 {
        return decimal_from_mantissa_exp(&mantissa, exponent);
    }
    let sign = if exponent >= 0 { '+' } else { '-' };
    let abs_exp = exponent.unsigned_abs();
    if abs_exp < 10 {
        format!("{}e{}0{}", mantissa, sign, abs_exp)
    } else {
        format!("{}e{}{}", mantissa, sign, abs_exp)
    }
}

fn strip_fractional_trailing_zeros(s: &str) -> String {
    if !s.contains('.') {
        return s.to_string();
    }
    let trimmed = s.trim_end_matches('0').trim_end_matches('.');
    trimmed.to_string()
}

/// Renders mantissa * 10^exp in plain decimal form, stripping trailing zeros
/// from any fractional part.
fn decimal_from_mantissa_exp(mantissa: &str, exp: i32) -> String {
    let (sign, mantissa) = if let Some(rest) = mantissa.strip_prefix('-') {
        ("-", rest)
    } else {
        ("", mantissa)
    };

    let (digits, integer_len) = match mantissa.find('.') {
        Some(idx) => {
            let mut s = String::with_capacity(mantissa.len() - 1);
            s.push_str(&mantissa[..idx]);
            s.push_str(&mantissa[idx + 1..]);
            (s, idx as i32)
        }
        None => (mantissa.to_string(), mantissa.len() as i32),
    };

    let new_integer_len = integer_len + exp;

    if new_integer_len <= 0 {
        let padding = "0".repeat((-new_integer_len) as usize);
        let raw = format!("0.{}{}", padding, digits);
        let trimmed = raw.trim_end_matches('0').trim_end_matches('.');
        return format!("{}{}", sign, trimmed);
    }
    if (new_integer_len as usize) >= digits.len() {
        let padding = "0".repeat((new_integer_len as usize) - digits.len());
        return format!("{}{}{}", sign, digits, padding);
    }
    let int_part = &digits[..new_integer_len as usize];
    let frac_part = digits[new_integer_len as usize..].trim_end_matches('0');
    if frac_part.is_empty() {
        format!("{}{}", sign, int_part)
    } else {
        format!("{}{}.{}", sign, int_part, frac_part)
    }
}

/// Validates that a float is encodable to JSON.
pub fn validate_f32(value: f32) -> Result<(), String> {
    if value.is_nan() {
        return Err("cannot encode NaN to JSON".to_string());
    }
    if value.is_infinite() {
        return Err("cannot encode Infinity to JSON".to_string());
    }
    Ok(())
}

/// Validates that a double is encodable to JSON.
pub fn validate_f64(value: f64) -> Result<(), String> {
    if value.is_nan() {
        return Err("cannot encode NaN to JSON".to_string());
    }
    if value.is_infinite() {
        return Err("cannot encode Infinity to JSON".to_string());
    }
    Ok(())
}

/// Encodes bytes to base64 string (RFC 4648 standard encoding).
pub fn encode_base64(data: &[u8]) -> String {
    general_purpose::STANDARD.encode(data)
}

/// Decodes a base64 string to bytes.
pub fn decode_base64(s: &str) -> Result<Vec<u8>, String> {
    general_purpose::STANDARD
        .decode(s)
        .map_err(|e| format!("invalid base64: {}", e))
}

/// Parses an i64 from a JSON value (string or number).
pub fn parse_i64_from_json(value: &serde_json::Value) -> Result<i64, String> {
    match value {
        serde_json::Value::String(s) => s
            .parse::<i64>()
            .map_err(|e| format!("failed to parse i64: {}", e)),
        serde_json::Value::Number(n) => n
            .as_i64()
            .ok_or_else(|| "number out of i64 range".to_string()),
        _ => Err("expected string or number for i64".to_string()),
    }
}

/// Parses a u64 from a JSON value (string or number).
pub fn parse_u64_from_json(value: &serde_json::Value) -> Result<u64, String> {
    match value {
        serde_json::Value::String(s) => s
            .parse::<u64>()
            .map_err(|e| format!("failed to parse u64: {}", e)),
        serde_json::Value::Number(n) => n
            .as_u64()
            .ok_or_else(|| "number out of u64 range".to_string()),
        _ => Err("expected string or number for u64".to_string()),
    }
}

/// Parses an i32 from a JSON value (string or number).
pub fn parse_i32_from_json(value: &serde_json::Value) -> Result<i32, String> {
    match value {
        serde_json::Value::String(s) => s
            .parse::<i32>()
            .map_err(|e| format!("failed to parse i32: {}", e)),
        serde_json::Value::Number(n) => n
            .as_i64()
            .ok_or_else(|| "not a number".to_string())
            .and_then(|v| i32::try_from(v).map_err(|e| format!("i32 overflow: {}", e))),
        _ => Err("expected string or number for i32".to_string()),
    }
}

/// Parses a u32 from a JSON value (string or number).
pub fn parse_u32_from_json(value: &serde_json::Value) -> Result<u32, String> {
    match value {
        serde_json::Value::String(s) => s
            .parse::<u32>()
            .map_err(|e| format!("failed to parse u32: {}", e)),
        serde_json::Value::Number(n) => n
            .as_u64()
            .ok_or_else(|| "not a number".to_string())
            .and_then(|v| u32::try_from(v).map_err(|e| format!("u32 overflow: {}", e))),
        _ => Err("expected string or number for u32".to_string()),
    }
}

/// Sorts map keys lexicographically by UTF-8 byte order.
/// This ensures deterministic JSON output for maps.
///
/// Accepts `&mut [String]` so the codegen-emitted callers passing
/// `&mut keys` (where `keys: Vec<String>`) auto-deref without requiring
/// the helper to own the vector type.
pub fn sort_map_keys_lexicographic(keys: &mut [String]) {
    keys.sort();
}

/// Escapes a string for safe inclusion in JSON.
/// Uses serde_json for correct escaping.
pub fn escape_json_string(s: &str) -> String {
    serde_json::to_string(s).unwrap_or_else(|_| "\"\"".to_string())
}

/// Helper to build JSON strings efficiently.
pub struct JsonBuilder {
    parts: Vec<String>,
}

impl JsonBuilder {
    pub fn new() -> Self {
        JsonBuilder { parts: Vec::new() }
    }

    pub fn push(&mut self, s: &str) {
        self.parts.push(s.to_string());
    }

    pub fn push_string(&mut self, s: String) {
        self.parts.push(s);
    }

    pub fn build(self) -> String {
        self.parts.join("")
    }
}

impl Default for JsonBuilder {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_i64_to_string() {
        assert_eq!(format_i64_to_string(0), "0");
        assert_eq!(format_i64_to_string(123), "123");
        assert_eq!(format_i64_to_string(-456), "-456");
        assert_eq!(format_i64_to_string(i64::MAX), "9223372036854775807");
        assert_eq!(format_i64_to_string(i64::MIN), "-9223372036854775808");
    }

    #[test]
    fn test_format_u64_to_string() {
        assert_eq!(format_u64_to_string(0), "0");
        assert_eq!(format_u64_to_string(123), "123");
        assert_eq!(format_u64_to_string(u64::MAX), "18446744073709551615");
    }

    #[test]
    #[allow(clippy::approx_constant)] // literals are reference values matched against Go output, not π/e
    fn test_format_f32() {
        // Reference values come from pkg/cramberry/json_test.go::TestFormatFloat32.
        // They must match Go's strconv.FormatFloat(v, 'g', 9, 32) byte-for-byte.
        assert_eq!(format_f32(0.0).unwrap(), "0");
        assert_eq!(format_f32(-0.0).unwrap(), "0");
        assert_eq!(format_f32(3.14159).unwrap(), "3.14159012");
        assert_eq!(format_f32(-2.71828).unwrap(), "-2.71828008");
        assert_eq!(format_f32(1.23e10).unwrap(), "1.23000003e+10");
        assert_eq!(format_f32(0.0000001).unwrap(), "1.00000001e-07");
        assert!(format_f32(f32::NAN).is_err());
        assert!(format_f32(f32::INFINITY).is_err());
        assert!(format_f32(f32::NEG_INFINITY).is_err());
    }

    #[test]
    #[allow(clippy::approx_constant)] // literals are reference values matched against Go output, not π/e
    fn test_format_f64() {
        // Reference values come from pkg/cramberry/json_test.go::TestFormatFloat64.
        assert_eq!(format_f64(0.0).unwrap(), "0");
        assert_eq!(format_f64(-0.0).unwrap(), "0");
        assert_eq!(format_f64(2.718281828459045).unwrap(), "2.7182818284590451");
        assert_eq!(
            format_f64(-3.141592653589793).unwrap(),
            "-3.1415926535897931"
        );
        assert_eq!(format_f64(1.23e100).unwrap(), "1.2300000000000001e+100");
        assert_eq!(format_f64(1e-100).unwrap(), "1e-100");
        assert_eq!(format_f64(1e20).unwrap(), "1e+20");
        assert_eq!(format_f64(1.5).unwrap(), "1.5");
        assert_eq!(format_f64(0.0001).unwrap(), "0.0001");
        assert_eq!(format_f64(1e-7).unwrap(), "9.9999999999999995e-08");
        assert!(format_f64(f64::NAN).is_err());
        assert!(format_f64(f64::INFINITY).is_err());
        assert!(format_f64(f64::NEG_INFINITY).is_err());
    }

    #[test]
    fn test_encode_decode_base64() {
        let data = vec![0xde, 0xad, 0xbe, 0xef];
        let encoded = encode_base64(&data);
        assert_eq!(encoded, "3q2+7w==");

        let decoded = decode_base64(&encoded).unwrap();
        assert_eq!(decoded, data);
    }

    #[test]
    fn test_parse_i64_from_json() {
        let value = serde_json::json!("123");
        assert_eq!(parse_i64_from_json(&value).unwrap(), 123);

        let value = serde_json::json!(456);
        assert_eq!(parse_i64_from_json(&value).unwrap(), 456);

        let value = serde_json::json!("-789");
        assert_eq!(parse_i64_from_json(&value).unwrap(), -789);
    }

    #[test]
    fn test_parse_u64_from_json() {
        let value = serde_json::json!("123");
        assert_eq!(parse_u64_from_json(&value).unwrap(), 123);

        let value = serde_json::json!(456);
        assert_eq!(parse_u64_from_json(&value).unwrap(), 456);
    }

    #[test]
    fn test_sort_map_keys_lexicographic() {
        let mut keys = vec![
            "zebra".to_string(),
            "apple".to_string(),
            "banana".to_string(),
        ];
        sort_map_keys_lexicographic(&mut keys);
        assert_eq!(keys, vec!["apple", "banana", "zebra"]);
    }

    #[test]
    fn test_escape_json_string() {
        assert_eq!(escape_json_string("hello"), "\"hello\"");
        assert_eq!(escape_json_string("say \"hello\""), "\"say \\\"hello\\\"\"");
        assert_eq!(escape_json_string("line1\nline2"), "\"line1\\nline2\"");
    }
}
